### DISCUSSION TITLE
Correct typing for selecting outer joined optional table?

### DISCUSSION THREAD

**Discussion by BaaaZen** [sqlalchemy/sqlalchemy#10173] [Usage Questions]
After migrating from SA 1.4 to SA 2.0 I ran into a problem creating a correct typed query with an outer join.
My statement looks like this:
```
#  current type of stmt: Select[Tuple[TableA, TableB]]
# required type of stmt: Select[Tuple[TableA, TableB|None]]
stmt = select(TableA, TableB) \
    .outerjoin(
        TableA.ref_tblB.and_(
            TableA.some_additional_conditions == TableB.another_column
        )
    )

#  current type of results: Sequence[Row[Tuple[TableA, TableB]]]
# required type of results: Sequence[Row[Tuple[TableA, TableB|None]]]   
results = sql.execute(stmt).all()
```

Because of the outer join the selected `TableB` is optional and may become `null/None`. What is the correct way to use the select statement to get the correct typing for `TableB`?

**Comment by CaselIT** [comment]
Hi,

I'm not sure if this can be supported at the moment.

It seems that also manually typing `stmt` as `Select[Tuple[TableA, TableB | None]]` does not work since it gives a type error, at least on pyright.

@zzzeek do you have ideas on how to automate this?

**Reply by zzzeek** [reply]
well it could never be automatic unless we made all the typing be optional for each class or SQL expression, in which case then it wouldn't work for the more typical case where the class /column is not nullable.  has to be an end-user manual intervention.

now what that intervention would *be* is likely some special alias of some kind.

if we can have the folks here play with this patch someone can produce a PR and some docs for this

```diff
diff --git a/lib/sqlalchemy/sql/_typing.py b/lib/sqlalchemy/sql/_typing.py
index b4f6fc71cf..86bdbe187c 100644
--- a/lib/sqlalchemy/sql/_typing.py
+++ b/lib/sqlalchemy/sql/_typing.py
@@ -147,13 +147,36 @@ sets; select(...), insert().returning(...), etc.
 
 """
 
+from typing import overload, Optional
+
+
+
 _TypedColumnClauseArgument = Union[
+    # note: TypedColumnsClauseRole seems to be overlapping with
+    # ExpressionElementRole and perhaps doesn't need to be here
     roles.TypedColumnsClauseRole[_T],
     "SQLCoreOperations[_T]",
     roles.ExpressionElementRole[_T],
     Type[_T],
 ]
 
+# this would have a nicer name and be documented, and also have tests
+@overload
+def Nullable(val: "SQLCoreOperations[_T]") -> "SQLCoreOperations[Optional[_T]]":
+    ...
+
+@overload
+def Nullable(val: roles.ExpressionElementRole[_T]) -> roles.ExpressionElementRole[Optional[_T]]:
+    ...
+
+@overload
+def Nullable(val: Type[_T]) -> Type[Optional[_T]]:
+    ...
+
+def Nullable(val: _TypedColumnClauseArgument[_T]) -> _TypedColumnClauseArgument[Optional[_T]]:
+    return val  # type: ignore
+
+
 _TP = TypeVar("_TP", bound=Tuple[Any, ...])
 
 _T0 = TypeVar("_T0", bound=Any)
```

here is a demo:

```py
from __future__ import annotations

from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import select
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session

# this would be in a nicer place
from sqlalchemy.sql._typing import Nullable


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[str]
    bs: Mapped[list[B]] = relationship("B")


class B(Base):
    __tablename__ = "b"
    id: Mapped[int] = mapped_column(primary_key=True)
    a_id: Mapped[int] = mapped_column(ForeignKey("a.id"))
    data: Mapped[str]



stmt1 = select(A, B)
stmt2 = select(A.data, B.data)
stmt3 = select(A, Nullable(B))
stmt4 = select(A.data, Nullable(B.data))


e = create_engine("sqlite://")
with Session(e) as sess:
    r1 = sess.execute(stmt1)

    r2 = sess.execute(stmt2)

    r3 = sess.execute(stmt3)

    r4 = sess.execute(stmt4)

    reveal_type(r1)
    reveal_type(r2)
    reveal_type(r3)
    reveal_type(r4)

```

here is the mypy output:

```
$ mypy test3.py 
test3.py:52: note: Revealed type is "sqlalchemy.engine.result.Result[tuple[test3.A, test3.B]]"
test3.py:53: note: Revealed type is "sqlalchemy.engine.result.Result[tuple[builtins.str, builtins.str]]"
test3.py:54: note: Revealed type is "sqlalchemy.engine.result.Result[tuple[test3.A, Union[test3.B, None]]]"
test3.py:55: note: Revealed type is "sqlalchemy.engine.result.Result[tuple[builtins.str, Union[builtins.str, None]]]"
Success: no issues found in 1 source file
```

**Reply by zzzeek** [reply]
noting also the above approach can be used in a program right now as it does not require any existing types be changed, only the addition of a new type that can be local to an application

**Reply by CaselIT** [reply]
I guess we could do that. an alternative may be to accept `select(Thing | None)`, but it's most likely not overly clear (it also uses typing at runtime, that's probably not the best thing to do)

**Reply by zzzeek** [reply]
`select(Thing | None)` is not a general solution, because for all the column / instrumented attribute objects, "Thing" is not a type.   So if `Thing` is a class, it has a chance of working but not for any of the other objects that are sent there, i.e. SQL expressions, which are more common for this case anyway.    so making it accept `Thing  | None` would mean we endorse something that only works in a small portion of the cases.

**Reply by CaselIT** [reply]
https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4793

where is the best place to document this?

**Reply by zzzeek** [reply]
somewhere in ORM Querying Guide, probably where it talks about different ways to `select()`, have a new small section re: typing

**Comment by CaselIT** [comment]
having a non nullable also makes sense, to be used for things like
```py
select(Model.some_nullable_value).where(Model.some_nullable_value.is_not(None))
```

**Reply by zzzeek** [reply]
in celebration of this (and as the owner of this project, as this is not something i want everyone doing all the time) I have made a meme to celebrate this idea

![image](https://github.com/sqlalchemy/sqlalchemy/assets/128223/96a27841-a669-4768-a499-e6812ed2f9a6)
